libqtile/widget/mpd2widget.py

```diff
@@ -165,28 +187,59 @@ def __getitem__(self, key):
 
         self.update(self.update_status)
 
-    def formatter(self, status, currentsong):
-        play_status = self.play_states[status['state']]
-
-        # Dirty hack to prevent keys conflict
-        currentsong['fulltime'] = currentsong['time']
-        del currentsong['time']
-
-        self.prepare_formatting(status, currentsong)
-        status.update(currentsong)
+    def formatter(self, status, current_song):
+        default = 'Undefined'
+        song_info = defaultdict(lambda: default)
```

Comment:
It doesn't look like anything here is necessarily needing the defaultdict, rather than a normal dict. What properties do you need to have added to keep the formatting from failing? Maybe you can conditionally add them at the end?

Reply:
I apparently have quite a number of corrupted tags in my library, all mp3 files from what I can tell.  MPD refuses to read the tags because of the presence of an invalid frame.  When this happens, the currentsong() method returns an empty dict.  Originally, I was using regex to fish the requested tags out of the format string, but I thought using a defaultdict was a better idea, as it would return a default value for any missing tag, without the overhead of a regular expression.

Reply:
I've thought of a few more edits, but I haven't pushed them to my fork yet.
1) For the sake of consistency, I moved the parameters 'prepare_status' and 'status_format' out of the constructor, and put them in the defaults dictionary instead.  I see that's the way the old mpd widget handled things.

2) I added an Idle state message that replaces the regular message when MPD is idle (e.g. status['state'] == 'stop' and current_song == {}) There's a secondary format string that I set to default to "{play_status} {idle_message} [{repeat}{random}{single}{consume}{updating_db}]"

3) I added a default formatting function ('all': lambda s: cgi.escape(s)) which prevents pango from throwing a fit if you use certain characters like "&" (I use this character because I'm using a special font to display the status symbols and the "shuffle" glyph is mapped to "&".  The character also shows up occasionally in artist information for songs. The 'all' function gets applied before any other user formatting.  I'm not sure if this is a great idea or not (partially why I haven't updated the fork) the presence of 'all' in the formatting function dict means that a user would then have to take that function's result into account with every other formatting function in the dictionary. This isn't difficult in the case of a simple escape sequence (I use ast.literal_eval to convert the escaped numbers back into numbers, but it's an extra step).  I did note that in my documentation, however.

let me know what you think.